docs/rfcs/blocks-in-widget-areas.md

```diff
@@ -0,0 +1,535 @@
+```
+
+## Storage
+
+WordPress currently stores which widgets belong to which widget areas using the `'sidebars_widgets'` site option. This is a serialized PHP array that maps widget area identifiers to a list of widget identifiers that are in that area.
+
+```php
+array(
+	'footer' => array(
+		'search-2',
+		'recent-posts-2',
+		'recent-comments-2',
+	),
+	'sidebar' => array(
+		'recent-posts-2',
+		'recent-comments-2',
+	),
+	'array_version' => 3,
+)
+```
+
+When a widget area stored this way is accessed via the REST API, widgets will be represented using `core/legacy-widget` blocks. Each `core/legacy-widget` block stores a reference to the existing widget (`identifier`), and, if it is a multi widget, the attributes that that instance has (`instance`).
+
+```
+GET /wp/v2/widget-areas/footer
+```
+
+```json
+{
+	"id": "footer",
+	...
+	"content": {
+		"raw": "
+			<!-- wp:legacy-widget {\"identifier\":\"search-2\",\"instance\":{\"title\":\"\"}} /-->
+			<!-- wp:legacy-widget {\"identifier\":\"recent-posts-2\",\"instance\":{\"title\":\"\",\"number\":5}} /-->
+			<!-- wp:legacy-widget {\"identifier\":\"recent-comments-2\",\"instance\":{\"title\":\"\",\"number\":5}} /-->
+		",
+		"rendered": ...,
+	}
+}
+```
+
+When a widget area is updated using the REST API, the storage format is _migrated_ and its `'array_version'` is set to `4`.
+
+After migration, the widget area identifier no longer maps to an array of widget identifiers. Instead, it maps to the ID of a `wp_area` post.
+
+`wp_area` is a new post type that denotes posts that contain block markup for a widget area. The `wp_area` post type is similar to the `wp_block` post type, except that it is marked as private and therefore not viewable by accessing `wp-admin/edit.php?post_type=wp_area`.
+
+Let's illustrate with an example. Say that we update the `'footer'` widget area to include a non-`core/legacy-widget` block.
+
+```
+PUT /wp/v2/widget-areas/footer
+
+{ "content": { "raw": "
+	<!-- wp:paragraph --><p>Hello there!</p><!-- /wp:paragraph -->
+	<!-- wp:legacy-widget {\"identifier\":\"search-2\",\"instance\":{\"title\":\"\"}} /-->
+	<!-- wp:legacy-widget {\"identifier\":\"recent-posts-2\",\"instance\":{\"title\":\"\",\"number\":5}} /-->
+	<!-- wp:legacy-widget {\"identifier\":\"recent-comments-2\",\"instance\":{\"title\":\"\",\"number\":5}} /-->
+" } }
+```
+
+```json
+{
+	"id": "footer",
+	...
+	"content": {
+		"raw": "
+			<!-- wp:paragraph --><p>Hello there!</p><!-- /wp:paragraph -->
+			<!-- wp:legacy-widget {\"identifier\":\"search-2\",\"instance\":{\"title\":\"\"}} /-->
+			<!-- wp:legacy-widget {\"identifier\":\"recent-posts-2\",\"instance\":{\"title\":\"\",\"number\":5}} /-->
+			<!-- wp:legacy-widget {\"identifier\":\"recent-comments-2\",\"instance\":{\"title\":\"\",\"number\":5}} /-->
+		",
+		"rendered": ...
+	}
+}
+```
+
+This triggers a migration. `'footer'` now points to a numeric ID instead of an array. Other widget areas are not affected.
+
+```php
+array(
+	'footer' => 123,
```

Comment:
What happens if I downgrade gutenberg or disable this new screen (which I think will be needed for at least some time).

Can't we instead do something like:

```php
array(
   'footer' => array( 'block-area-widget' ) // instance widget with the id 123
```

This is to say, instead of chaning the format, just keep the same and have a "hidden" widget called "block-area-widget" referencing a `wp_area` record.

The advantage is that all the php functions will work as expected, disabling this new behavior ensures a decent fallback.

Reply:
@youknowriad With the current approach @noisysocks is proposing, we can also hook into the sidebar areas and register a dynamic callback widget that displays the blocks saved the post that sidebar references as a widget in the widget area so we can be back-compatible.
In PR https://github.com/WordPress/gutenberg/pull/14251 @noisysocks is storing all the blocks as a string but then he dynamically parses the blocks and registers each one so each block appears in the widgets screen as a widget, here we could something similar.

I guess in the approach you are suggesting we would save the ids as instance attributes of the widget in the database. That approach has some performance disadvantages compared to this version.

We would first need to make a database query in the site options database, wait for the query result decode the widget instance and then make another query for the post id. In this approach, we just need to query the post id. So fewer queries to do.

Reply:
I'm hesitant about relying on "tricks" like this to ensure backwards compatibility. We're changing the data structure and shouldn't shy away from that. I'd rather that WordPress spits out a loud mess of PHP notices than for a subtle bug to be missed.

We can support disabling the new screen as @jorgefilipecosta nicely explains above.

We could support disabling the Gutenberg plugin by registering a [deactivation hook](https://developer.wordpress.org/reference/functions/register_deactivation_hook/) that does clean up and replaces the blocks in a widget area with a Text widget.

Supporting downgrading is trickier. Is it a requirement? I didn't think that WordPress ever supports being downgraded to an older version. Usually one restores from a backup instead.

Reply:
Can we document the proposed solution here in the RFC.

Reply:
Done in 7f59b59510c46419c7d26300ec93173c6a60600b.

Reply:
Reading this again in consideration of the "migration", I too had become curious if there could be some option where we don't need to change the structure. In my mind, I'd wondered if the structure could be left the same for existing widgets, and as new blocks content is added, doing so either as some custom widget, or perhaps even as the existing "Custom HTML" widget.

Considering some "Footer" with a search widget:

```php
array(
	'footer' => array(
		'search-2',
		// {"title":"Search for content:"}
	),
);
```

When presented in the widgets screen, the data would either be sent _as if_ it were blocks (`core/legacy-widget`) or as a widget, but interpreted on-the-fly to an equivalent legacy widget block in the client.

The interaction would feel the same as any other block editor. When ultimately saved, it might either send as blocks markup, or as the equivalent widgets data. In either case, what ends up being stored in the server aligns to this existing structure.

```php
array(
	'footer' => array(
		'search-2',
		// {"title":"Search for content:"}

		'custom_html-2', // or 'block_html-2', for semantic awareness
		// {"content":"<!-- wp:paragraph -->...<!-- /wp:paragraph -->\n\n<!-- wp:image -->...<!-- /wp:image -->"}
	),
);
```

This seems like the direction that #14251 had been heading with its "block widget". It's not clear the reasons it was pivoted toward an entirely new structure here. 

It does depend on a few factors, however:

- How simple is this "migration" to consider? It seems like a common point of contention for both timing and backwards-compatibility. If we could avoid it, that becomes moot.
- Is this idea of an "area" post type meant to satisfy some long-term objective (e.g. "content areas" from [9 Projects for 2019](https://make.wordpress.org/core/2018/12/08/9-priorities-for-2019/)), or merely as compatibility for widgets?
- Are we intent on abolishing the typical storage mechanism of widgets as an intentional consolidation of blocks markup as the storage mechanic?
   - Data consumed by the widgets screen would presumably either arrive from a widgets endpoint (which contradicts this goal to relegate widgets to legacy) or by some endpoint specific to this purpose (i.e. probably something like the "areas" just like what you're proposing). This could dismiss part of my worry; or not, if we grant that an areas endpoint should serve blocks markup, but that this be computed on the fly without changing the existing structure for widgets.

I'm not trying to shake things up dramatically at this stage, but in reviewing #15014 it had me considering _what even is an "area"_? And all of the effort we're putting in toward migrating storage.

Reply:
I do like the idea of minimising effort here, especially if we have a view to making widgets redundant in the medium to long term.

An approach involving a `WP_BlocksWidget` was the first I tried during prototyping. I ran into difficulties implementing the `PUT` endpoint as it requires an algorithm that doesn't unnecessarily create and delete existing widgets. Let me do some experimenting and get back to you.

Reply:
@jorgefilipecosta and I spoke about this and noted:

- Neither approach is really "simple". With CPTs the complexity is in migration and backwards compatibility. With HTML widgets the complexity is mostly in updating a widget area.
- Using CPTs has some scalability advantages noted in [#35669-core](https://core.trac.wordpress.org/ticket/35669).
- `wp_area` gives us flexibility in the future to implement "content areas". You could imagine that a "content area" is just a block that points to one or more `wp_area` IDs.

We agreed to continue with the `wp_area` approach. As with all things Gutenberg, though, experimentation and iteration is necessary and I'm comfortable revisiting this before merge if need be.